pkg/kubelet/logs/container_log_manager.go

```diff
@@ -171,20 +173,28 @@ func NewContainerLogManager(runtimeService internalapi.RuntimeService, osInterfa
 			MaxSize:  parsedMaxSize,
 			MaxFiles: maxFiles,
 		},
-		clock: clock.RealClock{},
-		mutex: sync.Mutex{},
+		clock:            clock.RealClock{},
+		mutex:            sync.Mutex{},
+		maxWorkers:       maxWorkers,
+		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kubelet_log_rotate_manager"),
+		monitoringPeriod: monitorDuration,
 	}, nil
 }
 
 // Start the container log manager.
 func (c *containerLogManager) Start() {
 	ctx := context.Background()
+	klog.InfoS("Initializing container log rotate workers", "workers", c.maxWorkers, "monitorPeriod", c.monitoringPeriod)
+	for i := 0; i <= c.maxWorkers; i++ {
+		worker := i + 1
+		go c.processQueueItems(worker)
+	}
 	// Start a goroutine periodically does container log rotation.
 	go wait.Forever(func() {
 		if err := c.rotateLogs(ctx); err != nil {
```

Comment:
Will this potentially create a large backlog of goroutines if the log rotate workers aren't able to keep up?

If I understand correctly, this will create a goroutine every 10s (currently), which will wait for the mutex lock to be available and do the work when it's available. This isn't a new problem and maybe it's not a big deal, but it would potentially be exacerbated if people are configuring to run more frequently that 10s.

Reply:
The wait loop won't kick off a new go-routine unless the current one is terminated.  So the large number of logs to be rotated should not cause an aggregation of the routines. Check the following snippet. Please do let me know if my understanding is wrong. 

```golang
package main

import (
	"fmt"
	"time"

	"k8s.io/apimachinery/pkg/util/wait"
)

func main() {
	wait.Forever(func() {
		fmt.Println("Inside")
		time.Sleep(5 * time.Second)
	}, 5*time.Millisecond)
}
```

Reply:
Makes sense